InventoryCatalog/Model/ResourceModel/StockStatusFilter.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Model\ResourceModel;
+
+use Magento\CatalogInventory\Api\StockConfigurationInterface;
+use Magento\Framework\DB\Select;
+use Magento\InventoryIndexer\Indexer\IndexStructure;
+use Magento\InventoryIndexer\Model\StockIndexTableNameResolverInterface;
+
+/**
+ * Generic in-stock status filter for multi stocks inventory
+ */
+class StockStatusFilter
+{
+    /**
+     * @var StockConfigurationInterface
+     */
+    private $stockConfiguration;
+    /**
+     * @var StockIndexTableNameResolverInterface
+     */
+    private $stockIndexTableProvider;
+
+    /**
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param StockIndexTableNameResolverInterface $stockIndexTableProvider
+     */
+    public function __construct(
+        StockConfigurationInterface $stockConfiguration,
+        StockIndexTableNameResolverInterface $stockIndexTableProvider
+    ) {
+        $this->stockConfiguration = $stockConfiguration;
+        $this->stockIndexTableProvider = $stockIndexTableProvider;
+    }
+
+    /**
+     * Add in-stock status constraint to the select for non default stock
+     *
+     * @param Select $select
+     * @param string $entityTable
+     * @param string $entityIdField
+     * @param string $stockStatusTableAlias
+     * @param int $stockId
+     * @param bool $force
+     * @return Select
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function execute(
+        Select $select,
+        string $entityTable,
+        string $entityIdField,
+        string $stockStatusTableAlias,
+        int $stockId,
+        bool $force = false
+    ): Select {
+        if ($force || !$this->stockConfiguration->isShowOutOfStock()) {
+            $productTableAlias = null;
+            $productTable = $select->getConnection()->getTableName('catalog_product_entity');
+            foreach ($select->getPart(Select::FROM) as $tableAlias => $part) {
```

Comment:
It's not a good idea to find table alias this way. Is it possible to pass it from the client code? Or split to the few filters for particular cases instead of implementing a generic one?